x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsIntegTests.java

```diff
@@ -1155,6 +1164,78 @@ private void assertSearchableSnapshotStats(String indexName, boolean cacheEnable
         }
     }
 
+    public void testMountingSnapshotLinksRefreshTaskAsChild() throws Exception {
+        final CyclicBarrier cyclicBarrier = new CyclicBarrier(2);
+
+        final String indexName = randomAlphaOfLength(10).toLowerCase(Locale.ROOT);
+        createAndPopulateIndex(indexName, indexSettingsNoReplicas(1).put(INDEX_SOFT_DELETES_SETTING.getKey(), true));
+
+        final String repositoryName = randomAlphaOfLength(10).toLowerCase(Locale.ROOT);
+        createRepository(repositoryName, "fs");
+
+        final SnapshotId snapshotOne = createSnapshot(repositoryName, "snapshot-1", List.of(indexName)).snapshotId();
+        assertAcked(indicesAdmin().prepareDelete(indexName));
+
+        // block the cluster state update thread, so we have the opportunity to inspect the mount/restore tasks' metadata
+        ClusterService clusterService = internalCluster().getCurrentMasterNodeInstance(ClusterService.class);
+        clusterService.submitUnbatchedStateUpdateTask("block master service", new ClusterStateUpdateTask() {
+            @Override
+            public ClusterState execute(ClusterState currentState) {
+                safeAwait(cyclicBarrier);   // notify test that we're blocked
+                safeAwait(cyclicBarrier);   // wait to be unblocked by test
+                return currentState;
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                throw new AssertionError("block master service", e);
+            }
+        });
+
+        // wait for master thread to be blocked
+        safeAwait(cyclicBarrier);
+
+        try {
+            final MountSearchableSnapshotRequest mountRequest = new MountSearchableSnapshotRequest(
+                indexName,
+                repositoryName,
+                snapshotOne.getName(),
+                indexName,
+                Settings.EMPTY,
+                Strings.EMPTY_ARRAY,
+                false,
+                MountSearchableSnapshotRequest.Storage.FULL_COPY
+            );
+
+            client().execute(MountSearchableSnapshotAction.INSTANCE, mountRequest, ActionListener.noop());
+            TaskInfo[] restoreSnapshotTask = new TaskInfo[1];
+            waitUntil(() -> {
+                restoreSnapshotTask[0] = getTaskForActionFromMaster(TransportRestoreSnapshotAction.TYPE.name());
+                return restoreSnapshotTask[0] != null;
+            });
```

Comment:
This still looks a bit janky, if there's a more suitable pattern about please advise. The idea was to just limit how long we poll the tasks API for in the event something goes wrong.

Reply:
I think just sticking an `assertBusy` around the whole thing would be fine. We only wait 10s if something goes wrong. It's true that any other failures in here will also cause `assertBusy` to retry but that's not really the point of this test, we have other tests to check these things.

Reply:
I agree waiting a bit here does not concern me. But if we really try to avoid polling, it is possible to achieve so by having a `ClusterStateObserver` that triggers and blocks when the index is assigned. We can be sure the tasks are already registered when that happens, i.e. something like the follows

```java
final var clusterStateObserver = new ClusterStateObserver(clusterService, null, logger, new ThreadContext(Settings.EMPTY));
final var indexAssignedBarrier = new CyclicBarrier(2);
clusterStateObserver.waitForNextChange(new ClusterStateObserver.Listener() {
    @Override
    public void onNewClusterState(ClusterState state) {
        safeAwait(indexAssignedBarrier);
        safeAwait(indexAssignedBarrier);
    }
    ...
}, state -> state.routingTable().hasIndex(indexName));
```

Reply:
Ah yeah nice idea.

Reply:
I went ahead with just using `assertBusy` for readability. To do this with the observer, I think we'd need to add the observer, submit the mount task, then wait on the observer which would mean it'd all need to be inline so probably not worth it given we're not too fussed about polling. Perhaps if there's value in it we could add something like `T ESTest#doSomethingThenWaitForIndexCreation(Supplier<T> action)` down the track.